### ISSUE TITLE
try-catch-finally

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#128]
There is an [editor's note](http://spec.xproc.org/master/head/xproc30/#ednote-finally) in the current version of the XProc 3.0 specs raising the question, whether a p:finally should be added to p:try and wether p:catch could be optional then.

I like this idea, because when the p:group raises an error which is not catched in this block, either because there is no p:catch at all or no p:catch for raised error code, I could do still some cleanup (e.g. remove created file etc.) before the error bubbles up.

So the suggestion is to introduce p:finally and to make p:catch optional (p:catch*) in the case there is a p:finally. The p:finally-block is ALWAYS executed, i.e.

- if p:group does not raise an error: after the last step of p:group.
- if p:group raises an error, that is catched: after the last step of the selected catch-block.
- if p:group raises an error which is not catched by a p:catch: after the step that raises the error, before the error bubble up.

Any thoughts, clarifications or objections?

**Comment by ndw** [comment]
There are some tricky parts: Can the finally block have any inputs? What inputs can it have? Does it know if the try failed? Does it know which catch fired?

**Comment by xml-project** [comment]
Yes, I think this is the point: We might easily sketch out a p:finally that tends to be useless.

The current version of the specs say there is no input or output port on p:finally, but I think it might be possible to say, that it has the same input ports as p:catch (including the error input port).
And (just thinking out loud): May be there is a second input port called "status" or something like this, with a document providing the information you mentioned: 
<c:status>
 <c:try-failed>true</c:try-failed>
 <c:error-catcher>(the name of the catching p:catch or empty, if not cached</c:error-catcher>
</c:status>

But may be, this gets to complex to be useful.

**Comment by xml-project** [comment]
Or: What about no input ports, but an XPath function (p:error-context()) visible in p:finally returning a map with the required context informations?

**Comment by ndw** [comment]
I wonder which is easier/better: a magic input or a magic function. We have the error port in catch, I guess we could send the error output to finally as well.

**Comment by ndw** [comment]
Does anyone have an actual use case for this?

**Comment by gimsieke** [comment]
A use case for `p:finally` at all, or a use case just for `p:finally` having any inputs, including the error port (which can give zero documents if `p:try` succeeded)?

I think the use cases for `p:finally` at all are limited in XProc anyway, since there aren’t that many “file open“ or “connection open“ operations that need to be unwinded after the operation succeeded or failed. Maybe there will even be fewer such operations since it is even less necessary to write stuff to disk in XProc 3.0.

But `p:finally` might be useful for some cases. In my view, it need not know any inputs or error states. This is how finally is designed in other languages, and everything that needs to know inputs and error state should be done by a `p:catch`. If there is an argument for putting repetetive code (for example, storing the error port output to a log file on disk before either consuming the error or raising it again) in `p:finally`, I’d argue that any such compound error handling behavior can be encapsulated in a single step which then can be called in each of the `p:catch` handlers, without too much verbosity.

So I’d say, let’s have a `p:finally` without any input since a) it may prove useful to some people and b) it seems not too complicated to implement. However, the environment (readable ports, in-scope-bindings) that is available in `p:group` should also be available in `p:finally`. 

Suppose you have this pipeline:
```xml
<p:variable name="tmpdir" select="concat(`file:/tmp/`, p:system-property('p:episode'))"/>
<!-- a zip file is on the default readable port -->
<p:try>
  <p:group>
    <their:unzip-to-disk dir="{$tmpdir}"/>
    <my:work-on-disk href="{$tmpdir}"/>
  </p:group>
  <p:catch code="my:ERR01">
    <p:message message="Operation on the files in '{$tmpdir}' failed."/>
  </p:catch>
  <p:finally>
    <p:delete-file recursive="true" href="{$tmpdir}"/>
  </p:finally>
</p:try>
```
then `p:finally` needs to know which directory to delete.
If the directory cannot be deleted because it couldn’t be created in `<their:unzip-to-disk/>` in the first place, the code in `p:finally` (unless `fail-on-error="false"` was set on `<p:delete-file/>`) will throw a new exception (`err:XD0021`?) that should bubble up before the uncaught exception of `<their:unzip-to-disk/>` (also an `err:XD0021`?) bubbles up.

**Comment by xml-project** [comment]
On the Aachen workshop we decided to have an optional "finally" 

- with arbitrary outputs (will be added to the outputs of the containing p:try) and 

- maybe the catch error port as error input port?

**Comment by ndw** [comment]
What's left to be done for this?

**Comment by xml-project** [comment]
See editorial note before 4.7.1.: The syntax has to be changed to either one or more "catch" or one "finally" after group. The error messages needs to be changed too. I make a PR this evening.

**Comment by xml-project** [comment]
Fixed with pr #209
